File: Crash it/AnimatedItem/AnimatedItem.swift
//
//  AnimatedItem.swift
//  Crash it
//
//  Created by Olivier Picard on 13/03/2018.
//  Copyright © 2018 Olivier Picard. All rights reserved.
//

import Foundation
import SpriteKit

class AnimatedItem: MovingItem {
    var loop:Int?
    var counterLoop:Int
    var latency: TimeInterval
    var textures:[SKTexture]
    var enableAnimation: Bool
    private var previous_frame: TimeInterval
    private var index_current_texture: Int
    private var delete_atEnd: Bool
    
    
    init(textureGroup textures: [SKTexture], numberOfLooping loop: Int?,
         size: CGSize, latency: TimeInterval, deleteAtEnd: Bool = true) {
        self.loop = loop
        self.counterLoop = 0
        self.textures = textures
        self.latency = latency
        self.previous_frame = 0
        self.index_current_texture = 0
        self.enableAnimation = true
        self.delete_atEnd = deleteAtEnd
        super.init(texture: textures[0], color: UIColor.white, size: size)
    }
    
    
    required init?(coder aDecoder: NSCoder) {
        self.textures = []
        self.counterLoop = 0
        self.latency = 0
        self.previous_frame = 0
        self.index_current_texture = 0
        self.enableAnimation = false
        self.delete_atEnd = true
        super.init(coder: aDecoder)
    }
    
    
    override func update(_ currentTime: TimeInterval) {
        super.update(currentTime)
        if(!enableAnimation) { return }
        if (self.loop ?? 0) > 0 && self.counterLoop >= (self.loop ?? 0) {
            if(self.delete_atEnd) {self.scene?.removeChildren(in: [self])}
            return
        }
        if currentTime - self.previous_frame < self.latency { return }
        
        if self.index_current_texture + 1 >= self.textures.count {
            self.index_current_texture = 0
            self.counterLoop += 1;
        } else { self.index_current_texture += 1 }
        
        self.texture = self.textures[self.index_current_texture]
        self.previous_frame = currentTime
        
    }
    
}

File: Android/app/src/main/java/com/olivierpicard/crachit/AnimatedItem/AnimatedItem.java
package com.olivierpicard.crachit.AnimatedItem;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import com.olivierpicard.crachit.Graphics.GSize;
import com.olivierpicard.crachit.MovingItem;
import com.olivierpicard.crachit.Tools;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by olivierpicard on 06/04/2018.
 */

public class AnimatedItem extends MovingItem {
    int loop;
    int counterLoop;
    long latency;
    List<Bitmap> bitmaps;
    boolean enableAnimation;
    private long previous_frame;
    private int index_current_texture;
    private boolean delete_atEnd;


    public AnimatedItem(List<Integer> bitmapRessouceIDs, int numberOfLooping,
                 GSize size, long latency, boolean deleteAtEnd) {
        super(bitmapRessouceIDs.get(0), size);
        bitmaps = new ArrayList<>();
        this.loop = numberOfLooping;
        this.counterLoop = 0;
        for(int bitmapRessouceID : bitmapRessouceIDs)
            bitmaps.add(BitmapFactory.decodeResource(Tools.resources, bitmapRessouceID));
        this.latency = latency;
        this.previous_frame = 0;
        this.index_current_texture = 0;
        this.enableAnimation = true;
        this.delete_atEnd = deleteAtEnd;
    }


    @Override
    public void update(long currentTime) {
        super.update(currentTime);
        if(!this.enableAnimation) return;
        if(this.loop > 0 && this.counterLoop >= this.loop
                && this.delete_atEnd && this.getScene() != null) {
            this.getScene().removeChild(this);
        }

        if(currentTime - this.previous_frame < this.latency) return;

        if(this.index_current_texture + 1 >= this.bitmaps.size()){
            this.index_current_texture = 0;
            this.counterLoop += 1;
        } else this.index_current_texture += 1;

        setBitmap(this.bitmaps.get(this.index_current_texture));
        this.previous_frame = currentTime;
    }


}
